cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -238,16 +238,62 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
+   * Inspects and builds plugin and application information for the given artifact, adding an additional set of
+   * plugin classes to the plugins found through inspection. This method is used when all plugin classes
+   * cannot be derived by inspecting the artifact but need to be explicitly set. This is true for 3rd party plugins
+   * like jdbc drivers.
+   *
+   * @param artifactId the id of the artifact to inspect and store
+   * @param artifactFile the artifact to inspect and store
+   * @param parentArtifacts artifacts the given artifact extends.
+   *                        If null, the given artifact does not extend another artifact
+   * @param additionalPlugins the set of additional plugin classes to add to the plugins found through inspection.
+   *                          If null, no additional plugin classes will be added
+   * @throws IOException if there was an exception reading from the artifact store
+   * @throws ArtifactRangeNotFoundException if none of the parent artifacts could be found
+   */
+  public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
+                                    @Nullable Set<ArtifactRange> parentArtifacts,
+                                    @Nullable Set<PluginClass> additionalPlugins)
+    throws IOException, ArtifactAlreadyExistsException, WriteConflictException,
+    InvalidArtifactException, ArtifactRangeNotFoundException {
+
+    parentArtifacts = parentArtifacts == null ? Collections.<ArtifactRange>emptySet() : parentArtifacts;
+    CloseableClassLoader parentClassLoader;
+    if (parentArtifacts.isEmpty()) {
+      parentClassLoader =
+        artifactClassLoaderFactory.createClassLoader(Locations.toLocation(artifactFile));
+    } else {
+      parentClassLoader = createParentClassLoader(artifactId, parentArtifacts);
+      validateParentSet(parentArtifacts);
     }
-    validateParentSet(parentArtifacts);
 
-    try (CloseableClassLoader parentClassLoader = createParentClassLoader(artifactId, parentArtifacts)) {
-      ArtifactClasses artifactClasses = artifactInspector.inspectArtifact(artifactId, artifactFile, parentClassLoader);
-      validatePluginSet(artifactClasses.getPlugins());
+    try {
+      ArtifactClasses artifactClasses = inspectArtifact(artifactId, artifactFile, additionalPlugins, parentClassLoader);
       ArtifactMeta meta = new ArtifactMeta(artifactClasses, parentArtifacts);
       return artifactStore.write(artifactId, meta, Files.newInputStreamSupplier(artifactFile));
+    } finally {
+      parentClassLoader.close();
+    }
+  }
+
+  private ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile,
+                                          @Nullable Set<PluginClass> additionalPlugins,
+                                          ClassLoader parentClassLoader) throws IOException, InvalidArtifactException {
+    ArtifactClasses artifactClasses = artifactInspector.inspectArtifact(artifactId, artifactFile, parentClassLoader);
```

Comment:
yeah was debating whether to do only or addition.  The reason decided on in addition is that I could think of a case where you've included a 3rd party plugin with a bunch of other explicitly annotated plugins, in which case you would want it to be in addition rather than only. But could not think of a situation where you would want 'only' when there are annotated plugins in the artifact.